enhancements/two-nodes-openshift/2no.md

```diff
@@ -0,0 +1,486 @@
+* As an enterprise running workloads on a minimal OpenShift footprint, I want to minimize time-to-recovery and data loss for my workloads when a node fails.
+
+### Goals
+
+* Provide a two-node control-plane for physical hardware that is resilient to a node-level failure for either node
+* Provide a transparent installation experience that starts with exactly 2 blank physical nodes, and ends with a fault-tolerant two-node cluster
+* Prevent both data corruption and divergent datasets in etcd
+* Minimize recovery-caused unavailability. Eg. by avoiding fencing loops, wherein each node powers cycles its peer after booting, reducing the cluster's availability.
+* Recover the API server in less than 120s, as measured by the surviving node's detection of a failure
+* Minimize any differences to existing OpenShift topologies
+* Avoid any decisions that would prevent future implementation and support for upgrade/downgrade paths between two-node and traditional architectures
+* Provide an OpenShift cluster experience that is similar to that of a 3-node hyperconverged cluster but with 2 nodes
+
+### Non-Goals
+
+* Workload resilience - see related [Pre-DRAFT enhancement](https://docs.google.com/document/d/1TDU_4I4LP6Z9_HugeC-kaQ297YvqVJQhBs06lRIC9m8/edit)
+* Resilient storage - see future enhancement
+* Support for platforms other than bare metal including automated CI testing
+* Support for other topologies (eg. hypershift)
+* Support disconnected cluster installation
+* Adding worker nodes
+* Creation of RHEL-HA events and metrics for consumption by the OpenShift monitoring stack (Deferred to post-MVP)
+* Supporting upgrade/downgrade paths between two-node and other architectures (for initial release)
+
+## Proposal
+
+Use the RHEL-HA stack (Corosync, and Pacemaker), which has been used to deliver supported 2-node cluster experiences for multiple decades, to manage cri-o, kubelet, and the etcd daemon.
+etcd will run as as a voting member on both nodes.
+We will take advantage of RHEL-HA's native support for systemd and re-use the standard cri-o and kubelet units, as well as create a new Open Cluster Framework (OCF) script for etcd.
+The existing startup order of cri-o, then kubelet, then etcd will be preserved.
+The `etcdctl`, `etcd-metrics`, and `etcd-readyz` containers will remain part of the static pod, the contents of which remain under the exclusive control of the Cluster Etcd Operator (CEO).
+
+Use RedFish compatible Baseboard Management Controllers (BMCs) as our primary mechanism to power off (fence) an unreachable peer and ensure that it can do no harm while the remaining node continues.
+
+Upon a peer failure, the RHEL-HA components on the survivor will fence the peer and use the OCF script to restart etcd as a new cluster-of-one.
+
+Upon a network failure, the RHEL-HA components ensure that exactly one node will survive, fence its peer, and use the OCF script to restart etcd as a new cluster-of-one.
+
```

Comment:
This would follow the same response stated in the sentence above

> use the OCF script to restart etcd as a new cluster-of-one.

however, I don't think there is the need to _fence_ the peer in this case

Reply:
So is the take-away here that we are implementing our own kind of health probe in the OCF script?

Reply:
yes. OCF script already has a monitor function implemented, which at the moment verifies the member list, for example. We must extend the checks to other aspects of etcd

Reply:
I'll update the EP

Reply:
updated https://github.com/openshift/enhancements/pull/1675/commits/e28d72524044a0e76c9ec7a3e06717f09c91a82e